### PR TITLE
[pyspark] fix empty data issue when constructing DMatrix

### PR DESCRIPTION
**Pull request by wbo4958**
1. Fix empty data issue when user specifies the validation col
2. Fix the empty partition issues

To fix https://github.com/dmlc/xgboost/issues/8221

### CODE REVIEW

File: python-package/xgboost/spark/data.py [hunk 1]
```diff
@@ -247,9 +252,14 @@ def make(values: Dict[str, List[np.ndarray]], kwargs: Dict[str, Any]) -> DMatrix
...
         dtrain = DeviceQuantileDMatrix(it, **kwargs)
 
-    dvalid = make(valid_data, kwargs) if len(valid_data) != 0 else None
+    # Using has_validation_col here to indicate if there is validation col
+    # instead of getting it from iterator, since the iterator may be empty
+    # in some special case. That is to say, we must ensure every worker
+    # construct DMatrix even there is no any data since we need to ensure every
+    # worker do the AllReduce when constructing DMatrix, or else it may hang
+    # forever.
+    dvalid = make(valid_data, kwargs) if has_validation_col else None
```
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:261]: Good fix!

**Review by WeichenXu123 — APPROVED** [review]

**Review by trivialfis — APPROVED** [review]
Thank you for the fix. Please resolve the CI errors.

**Comment by hcho3** [comment]
Re-running the CI, since we recently transitioned to another CI system

**Comment by wbo4958** [comment]
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-jar-plugin:3.0.2:jar (default-jar) on project xgboost4j-gpu_2.12: Execution default-jar of goal org.apache.maven.plugins:maven-jar-plugin:3.0.2:jar failed: Plugin org.apache.maven.plugins:maven-jar-plugin:3.0.2 or one of its dependencies could not be resolved: Could not transfer artifact org.codehaus.plexus:plexus-component-annotations:jar:1.6 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]` Seems connection reset?

**Comment by wbo4958** [comment]
@hcho3 could you help to re-trigger this build?

**Comment by hcho3** [comment]
Done

**Comment by trivialfis** [comment]
It seems the test is failing: https://buildkite.com/xgboost/xgboost-ci/builds/127#01833e0b-06cf-4fd3-b29c-dc430e6cff04 .

**Review by WeichenXu123 — APPROVED** [review]

**Comment by wbo4958** [comment]
@hcho3 could you help to trigger CI?

**Comment by trivialfis** [comment]
Started the CI.

**Pull request merged**